archimedes/clients/dashboard.py

```diff
@@ -59,18 +60,20 @@ def export_dashboard(self, dashboard_id):
         """
         url = urijoin(self.base_url, self.API_DASHBOARDS_URL, self.API_EXPORT_COMMAND + '?dashboard=' + dashboard_id)
 
-        dashboard = None
         try:
             dashboard = self.fetch(url)
 
             errors = [obj for obj in dashboard['objects'] if 'error' in obj]
             if errors:
                 msg = errors[0]['error']['message'].lower()
-                logger.error("Impossible to export dashboard with id %s, %s", dashboard_id, msg)
-                dashboard = None
+                cause = "Impossible to export dashboard with id %s, %s" % (dashboard_id, msg)
+                logger.error(cause)
+                raise ExportError(cause=cause)
         except requests.exceptions.HTTPError as error:
             if error.response.status_code == 400:
-                logger.error("Impossible to export dashboard with id %s", dashboard_id)
+                cause = "Impossible to export dashboard with id %s" % dashboard_id
```

Comment:
Why the server would reply with a 400 error? That's a Bad Request Error, so it means the client (this code) is sending something wrong. I wonder what Archimedes would do wrong. If it does, probably it shouldn't be managed here.

Reply:
The error pops up when trying to export a dashboard, previously uploaded with a missing visualization. Conversely, when the index pattern is missing, the error is incapsulated in the response (https://github.com/Bitergia/archimedes/blob/master/archimedes/clients/dashboard.py#L68)

To replicate this behaviour you can use the following script:
```
import requests
    from archimedes.clients.dashboard import Dashboard
    from archimedes.errors import ExportError

    all_objects = {
          "objects": [
            {
              "id": "80b956f0-b2cd-11e8-ad8e-85441f0c2e5c",
              "type": "visualization",
              "updated_at": "2018-09-07T18:40:33.247Z",
              "version": 1,
              "attributes": {
                "title": "Count Example",
                "visState": "{\"title\":\"Count Example\",\"type\":\"metric\",\"params\":{\"addTooltip\":true,\"addLegend\":false,\"type\":\"metric\",\"metric\":{\"percentageMode\":false,\"useRanges\":false,\"colorSchema\":\"Green to Red\",\"metricColorMode\":\"None\",\"colorsRange\":[{\"from\":0,\"to\":10000}],\"labels\":{\"show\":true},\"invertColors\":false,\"style\":{\"bgFill\":\"#000\",\"bgColor\":false,\"labelColor\":false,\"subText\":\"\",\"fontSize\":60}}},\"aggs\":[{\"id\":\"1\",\"enabled\":true,\"type\":\"count\",\"schema\":\"metric\",\"params\":{}}]}",
                "uiStateJSON": "{}",
                "description": "",
                "version": 1,
                "kibanaSavedObjectMeta": {
                  "searchSourceJSON": "{\"index\":\"90943e30-9a47-11e8-b64d-95841ca0b247\",\"query\":{\"query\":\"\",\"language\":\"lucene\"},\"filter\":[]}"
                }
              }
            },
            {
              "id": "90943e30-9a47-11e8-b64d-95841ca0b247",
              "type": "index-pattern",
              "updated_at": "2018-09-07T18:39:47.683Z",
              "version": 1,
              "attributes": {
                "title": "kibana_sample_data_logs",
                "timeFieldName": "timestamp",
                "fields": "<truncated for example>",
                "fieldFormatMap": "{\"hour_of_day\":{}}"
              }
            },
            {
              "id": "942dcef0-b2cd-11e8-ad8e-85441f0c2e5c",
              "type": "dashboard",
              "updated_at": "2018-09-07T18:41:05.887Z",
              "version": 1,
              "attributes": {
                "title": "Example Dashboard",
                "hits": 0,
                "description": "",
                "panelsJSON": "[{\"gridData\":{\"w\":24,\"h\":15,\"x\":0,\"y\":0,\"i\":\"1\"},\"version\":\"7.0.0-alpha1\",\"panelIndex\":\"1\",\"type\":\"visualization\",\"id\":\"80b956f0-b2cd-11e8-ad8e-85441f0c2e5c\",\"embeddableConfig\":{}}]",
                "optionsJSON": "{\"useMargins\":true,\"hidePanelTitles\":false}",
                "version": 1,
                "timeRestore": False,
                "kibanaSavedObjectMeta": {
                  "searchSourceJSON": "{\"query\":{\"query\":\"\",\"language\":\"lucene\"},\"filter\":[]}"
                }
              }
            }
          ]
        }

    no_ip = {
        "objects": [
            {
                "id": "80b956f0-b2cd-11e8-ad8e-85441f0c2e5c",
                "type": "visualization",
                "updated_at": "2018-09-07T18:40:33.247Z",
                "version": 1,
                "attributes": {
                    "title": "Count Example",
                    "visState": "{\"title\":\"Count Example\",\"type\":\"metric\",\"params\":{\"addTooltip\":true,\"addLegend\":false,\"type\":\"metric\",\"metric\":{\"percentageMode\":false,\"useRanges\":false,\"colorSchema\":\"Green to Red\",\"metricColorMode\":\"None\",\"colorsRange\":[{\"from\":0,\"to\":10000}],\"labels\":{\"show\":true},\"invertColors\":false,\"style\":{\"bgFill\":\"#000\",\"bgColor\":false,\"labelColor\":false,\"subText\":\"\",\"fontSize\":60}}},\"aggs\":[{\"id\":\"1\",\"enabled\":true,\"type\":\"count\",\"schema\":\"metric\",\"params\":{}}]}",
                    "uiStateJSON": "{}",
                    "description": "",
                    "version": 1,
                    "kibanaSavedObjectMeta": {
                        "searchSourceJSON": "{\"index\":\"90943e30-9a47-11e8-b64d-95841ca0b247\",\"query\":{\"query\":\"\",\"language\":\"lucene\"},\"filter\":[]}"
                    }
                }
            },
            {
                "id": "942dcef0-b2cd-11e8-ad8e-85441f0c2e5c",
                "type": "dashboard",
                "updated_at": "2018-09-07T18:41:05.887Z",
                "version": 1,
                "attributes": {
                    "title": "Example Dashboard",
                    "hits": 0,
                    "description": "",
                    "panelsJSON": "[{\"gridData\":{\"w\":24,\"h\":15,\"x\":0,\"y\":0,\"i\":\"1\"},\"version\":\"7.0.0-alpha1\",\"panelIndex\":\"1\",\"type\":\"visualization\",\"id\":\"80b956f0-b2cd-11e8-ad8e-85441f0c2e5c\",\"embeddableConfig\":{}}]",
                    "optionsJSON": "{\"useMargins\":true,\"hidePanelTitles\":false}",
                    "version": 1,
                    "timeRestore": False,
                    "kibanaSavedObjectMeta": {
                        "searchSourceJSON": "{\"query\":{\"query\":\"\",\"language\":\"lucene\"},\"filter\":[]}"
                    }
                }
            }
        ]
    }

    no_viz = {
        "objects": [
            {
                "id": "90943e30-9a47-11e8-b64d-95841ca0b247",
                "type": "index-pattern",
                "updated_at": "2018-09-07T18:39:47.683Z",
                "version": 1,
                "attributes": {
                    "title": "kibana_sample_data_logs",
                    "timeFieldName": "timestamp",
                    "fields": "<truncated for example>",
                    "fieldFormatMap": "{\"hour_of_day\":{}}"
                }
            },
            {
                "id": "942dcef0-b2cd-11e8-ad8e-85441f0c2e5c",
                "type": "dashboard",
                "updated_at": "2018-09-07T18:41:05.887Z",
                "version": 1,
                "attributes": {
                    "title": "Example Dashboard",
                    "hits": 0,
                    "description": "",
                    "panelsJSON": "[{\"gridData\":{\"w\":24,\"h\":15,\"x\":0,\"y\":0,\"i\":\"1\"},\"version\":\"7.0.0-alpha1\",\"panelIndex\":\"1\",\"type\":\"visualization\",\"id\":\"80b956f0-b2cd-11e8-ad8e-85441f0c2e5c\",\"embeddableConfig\":{}}]",
                    "optionsJSON": "{\"useMargins\":true,\"hidePanelTitles\":false}",
                    "version": 1,
                    "timeRestore": False,
                    "kibanaSavedObjectMeta": {
                        "searchSourceJSON": "{\"query\":{\"query\":\"\",\"language\":\"lucene\"},\"filter\":[]}"
                    }
                }
            }
        ]
    }

    d = Dashboard('http://admin:admin@localhost:5601')

    # no errors
    d.import_objects(all_objects)
    d.export_dashboard("942dcef0-b2cd-11e8-ad8e-85441f0c2e5c")

    # delete .kibana
    requests.delete('https://admin:admin@localhost:9200/.kibana', verify=False)

    # error incapsulated
    d.import_objects(no_ip)
    try:
        d.export_dashboard("942dcef0-b2cd-11e8-ad8e-85441f0c2e5c")
    except ExportError:
        pass

    requests.delete('https://admin:admin@localhost:9200/.kibana', verify=False)

    # error in http
    d.import_objects(no_viz)
    try:
        d.export_dashboard("942dcef0-b2cd-11e8-ad8e-85441f0c2e5c")
    except ExportError:
        pass

    requests.delete('https://admin:admin@localhost:9200/.kibana', verify=False)
```

Reply:
Understood.